Feature/gltf2 importer

johnmaf opened the pull request.

dhritzkiv committed:
Duplicate gltfImporter as gltf2Importer; Include glTF2 importer in CMake List

dhritzkiv committed:
Update glTF2 Asset to use indexes

dhritzkiv committed:
Disambiguate Get methods

dhritzkiv committed:
Cache retrieved items via an original index map

dhritzkiv committed:
Start reading pbr materials

dhritzkiv committed:
temporarily disable gltf exporting of animations and skins

dhritzkiv committed:
Start managing and importing gltf2 pbr materials

dhritzkiv committed:
check in gltf2 models to test directory

dhritzkiv committed:
Working read, import, export, and write of gltf2 (pbr) material

dhritzkiv committed:
Add DamagedHelmet test model

dhritzkiv committed:
Remove Suzanne model

dhritzkiv committed:
specularGlossiness example

dhritzkiv committed:
Use `!ObjectEmpty()` vs. `MemberCount() > 0`

dhritzkiv committed:
Remove redundant function

dhritzkiv committed:
Start removing materials common, and adding pbrSpecularGlossiness

dhritzkiv committed:
Merge branch 'feature/gltf2-importer' of github.com:Matter-and-Form/assimp into feature/gltf2-importer

dhritzkiv committed:
Remove Light, Technique references

dhritzkiv committed:
Changes to GLTF2 materials

pbrMetallicRoughness and pbrSpecularGlossiness as structs; persist textureinfo properties from start to finish; persist pbrSpecularGlossiness (via extensionsUsed) usage from start to finish

dhritzkiv committed:
Remove OPEN3DGC and compression references

dhritzkiv committed:
mention pbrSpecularGlossiness support

dhritzkiv committed:
Make sure `on` flag for specularGlossiness is being persisted

dhritzkiv committed:
Write specularGlossiness textures on the specularGlossiness object

dhritzkiv committed:
reenable animation and skins exports

Currently incorrect, however. May need to be removed

dhritzkiv committed:
store node mesh vs. meshes

glTF nodes can only hold one mesh. this simply assigns to and check’s a Node’s Mesh

dhritzkiv committed:
Export extensions last

dhritzkiv committed:
Only export byteStride if not 0

dhritzkiv committed:
Formatting

dhritzkiv committed:
Remove ExportExtensions from glTF2 Exporter; Don’t store extensions used on rootNode metadata

exporting extensions happens in AssetWriter

dhritzkiv committed:
Fix unused CopyValue

dhritzkiv committed:
Set default values on Sampler

dhritzkiv committed:
Sampler improvements; Add new LazyDict method

Samplers are now imported into assimp internal format from textures;
Samplers have defaults as per spec;
Sampler enums are strongly typed with UNSET values;
Samplers are exported properly;
Sampler filters are exported as well;
Samplers are re-used across textures on export
Default sampler values are not written

dhritzkiv committed:
Export material names properly

dhritzkiv committed:
Remove Box MaterialsCommon models

As they are invalid

dhritzkiv committed:
Re-include gltf (1) importer

Currently doesn’t do gltf1 importing

dhritzkiv commented on code/glTF2AssetWriter.inl:
```diff
@@ -196,51 +206,158 @@ namespace glTF2 {
 
     namespace {
-        inline void WriteTex(Value& obj, Ref<Texture> texture, const char* propName, MemoryPoolAllocator<>& al)
+        inline void SetTexBasic(TextureInfo t, Value& tex, MemoryPoolAllocator<>& al)
```
This method I created to be used in the various forms of `WriteTex`. Is this the right way to create a reusable function that's used only within the `WriteTex` functions?

johnmaf replied:
Yeah, that's a good trick!

dhritzkiv commented on code/glTF2AssetWriter.inl:
```diff
@@ -196,51 +206,158 @@ namespace glTF2 {
+            WriteTex(pbrMetallicRoughness, m.pbrMetallicRoughness.metallicRoughnessTexture, "metallicRoughnessTexture", w.mAl);
+
+            //@TODO: define this as a constant?
+            vec4 defaultEmissiveFactor = {1, 1, 1, 1};
```
If I wanted to define these default values (used for comparing against the incoming values), where would be the best place to define them?

johnmaf replied:
If they need to be accessible outside that file, you might consider putting them inside `glTF2AsserWriter.h`, and defining them as `static const vec4 ...`, which means they'll only be defined once in the life of the program--that is, they're `static`--and they'll be read-only (`const`).

dhritzkiv commented on code/glTF2AssetWriter.inl:
```diff
@@ -196,51 +206,158 @@ namespace glTF2 {
+        Value exts;
+        exts.SetObject();
+
+        if (m.pbrSpecularGlossiness.on) {
```
Is `.on` a valid form of flagging whether pbrSpecularGlossiness is enabled? Or should I use/instantiate a pointer, and check whether or not it is null?

johnmaf replied:
Unfortunately C++ doesn't have an idiomatic way to do this--what it really needs is an option type. My personal preference, which doesn't seem to be the norm, is to use a `std::unique_ptr<T>`, which is an object that wraps a `T*`--pointer to `T`--and automatically `delete`s it, so you don't have to worry so much about memory leaks. Another option is to have a `pbrSpecularGlossinessOn` flag on the material itself. Your solution seems at least as clean as those two, so it's up to you.

dhritzkiv replied:
Your preference seems like a good idea, but is a completely foreign concept to me. If you think it's better, feel free to make the changes.

johnmaf replied:
Nah, I think what you've done is much simpler and cleaner. Using pointers as an option type is a bit of a non-obvious abuse of notation, even though they have the same semantics.

johnmaf replied:
Ooh, just noticed that there's a simple `Nullable` struct already defined in `glTF2Asset.h`: https://github.com/Matter-and-Form/assimp/blob/26c3bbfc0ffcae368c5eb9229ee67c35875a6065/code/glTF2Asset.h#L372.
So maybe you could make the `pbrSpecularGlossiness` a `Nullable<PbrSpecularGlossiness>` and omit the `on` property?

dhritzkiv replied:
How do I enable/disable (null / not-null) it in the code?

johnmaf replied:
There's an `isPresent` flag on the `Nullable`. If you use the constructor that takes no arguments, it's automatically set to `false`; if you pass in a value to the constructor, it's `true`.

dhritzkiv commented on code/glTF2Importer.cpp:
```diff
@@ -0,0 +1,651 @@
+        SetMaterialColorProperty(r, mat.emissiveFactor, aimat, AI_MATKEY_COLOR_EMISSIVE);
+
+        aimat->AddProperty(&mat.doubleSided, 1, AI_MATKEY_TWOSIDED);
+        aimat->AddProperty(&mat.alphaMode, 1, "$mat.gltf.alphaMode");
```
Should I write these custom property names inline (as-is), or is it better to `DEFINE` them somewhere else (like the GLTF2Asset file) for reusability/consistency?

johnmaf replied:
If you need to use the same string twice in the source code, I'd say you should `#define` it somewhere with a carefully "namespaced" name, like `AI_GLTF_MATKEY_ALPHAMODE`. Alternatively, you could define it as a `static const` variable.

dhritzkiv commented on code/glTF2Asset.h:
```diff
@@ -232,7 +230,7 @@ namespace glTF2
                 return 1;
             default:
-                throw DeadlyImportError("GLTF: Unsupported Component Type "+t);
+                throw DeadlyImportError("GLTF: Unsupported Component Type " + std::to_string(t));
```
Clang was warning about  simply appending `t` to the string with the + sign. Is this better?

johnmaf replied:
Yes. The literal `"GLTF: Unsupported Component Type "` is a C-style string, which is a character pointer (`char *`) with the convention that the end of the string is indicated by a null byte. These are sometimes called "null-terminated strings".

So, suppose the incoming `ComponentType` `t` happens to be `ComponentType_BYTE`, which we've specified to have an underlying integer value of `5120`. Then the expression `"my C-style string" + t` consists in adding an `enum ComponentType` to a `char *`, and the result will be of the form `<address of my c-style string> + 5120`, which will cause the `DeadlyImportError` to be initialized with whatever happens to be in memory at that location, reading until it finds a null byte.

If, however, we explicitly convert `t` to a `std::string`--an object provided by the C++ standard library in the `<string>` header--as in the expression `"my c-style string" + std::to_string(t)`, the compiler uses the `+` operator overloaded by the `std::string` class, and the result is a `std::string` of the form `"my c-style string<whatever's in t>"`.

dhritzkiv commented on code/glTF2Asset.h:
```diff
@@ -244,15 +242,17 @@ namespace glTF2
 
     //! Values for the Sampler::magFilter field
-    enum SamplerMagFilter
+    enum class SamplerMagFilter: unsigned int
```
I "had" to use strongly typed enums as I needed a way to set a unset state; simply checking an uninitialized enum value was undefined behaviour. Is this okay? Will assimp maintainers be okay with this form? Or is it overkill?

johnmaf replied:
I personally prefer `enum class` for various reasons, but I think you could still do this with a regular `enum`, no? It's very common to define something like the following:
```c++
enum Something {
    Something_UNKOWN = 0,
    Something_ELSE = 1
}
```
Also, as a stylistic matter, one of the nice things about `enum class`es is that their members are scoped, so you could do something like:
```c++
    enum class SamplerMagFilter: unsigned int
    {
        UNSET = 0,
        Nearest = 9728,
        Linear = 9729
    };
```
That is, omitting the `SamplerMagFilter_` prefix, which is only used in the regular `enum` because its members aren't scoped, which causes problems in situations like the following:
```c++
// my_code.h
enum MyBool {
    TRUE,
    FALSE
}

// some_library.h
enum LibraryBool {
    TRUE,
    FALSE
}

// main.cpp
#include "my_code.h"
#include "some_library.h"
// -> compiler complains about re-declaration of `TRUE` and `FALSE`
```

dhritzkiv replied:
Ah, right. I think that's why I ultimately chose "enum class" because they had the side-effect of being scoped, so I could re-use `UNSET` across different enums

dhritzkiv commented on code/glTF2Asset.h:
```diff
@@ -629,7 +631,7 @@ namespace glTF2
                 float znear; //! The floating-point distance to the near clipping plane. (required)
             } ortographic;
-        };
+        } cameraProperties;
```
Is this the right way to name a union?

johnmaf replied:
The statement `union { ... } myName;` is a variable declaration, as are `struct { ... } orthographic` and `struct { ... } perspective`. So, in particular, you can't ever create another variable of that `union` type, or of either of the `struct`s you defined in the union, since you haven't given them a type name: they're all three of them anonymous types. If that's what you're looking for, this is perfectly valid syntax.

If, however, you were looking to define a `struct` with the type name `perspective`, you would do something like the following:
```c++
typedef struct perspective {
} perspective;
```
Which, read from the inside out, declares a type `struct perspective`, with whatever members you care to put inside the braces, and defines a type-alias `perspective` to refer to this type, omitting the `struct` portion.

dhritzkiv replied:
I'm… actually not sure what I want in this case lol. I think it was more a fix for a clang warning about anonymous types within anonymous types

johnmaf replied:
Hmm...well, do you ever want to be able to do something like `perspective myCamera { ... }; perspective.fov = 90;`? That currently isn't possible. Similarly, it wouldn't be possible to create a `cameraProperties`; the only one that can ever exist is the one you created and named `cameraProperties`.

dhritzkiv commented on code/glTF2Asset.h:
```diff
@@ -670,41 +672,66 @@ namespace glTF2
         Ref<Texture> texture;
-        vec4 color;
+        unsigned int index;
+        unsigned int texCoord = 0;
```
I noticed other structs initialized values in a `SetDefault()` method. Is there anything invalid about setting the default value inline like this?

johnmaf replied:
Nope, not that I know of.

dhritzkiv commented on code/glTF2Asset.inl:
```diff
@@ -187,47 +194,75 @@ inline void LazyDict<T>::DetachFromDocument()
+        throw DeadlyImportError("GLTF: Field is not an array \"" + std::string(mDictId) + "\"");
+    }
+
+    Value& obj = mDict->operator[](i);
```
Is this the right way to access the value at `i`? Or is this unsafe/prone-to-failure? I noticed that iterators are used everywhere else (for safety checks?)

johnmaf replied:
Function declarations like `operator[]`, `operator+`, etc. are used to overload the standard operators in C++. For example, `std::string` overloads the `+` operator to allow concatenating strings to `char *`s, or to other strings. In this case, rapidjson's `Value` type overloads the `[]` operator, to allow you to do things like `myValue[i]`, instead of `myValue.GetObjectAtIndex(i)`.

So one thing I'd do here is change this line to `Value &obj = (*mDict)[i];`, which is exactly equivalent, but more idiomatic, since it uses the operator directly instead of calling the `operator[]` member function.

That said, you may be right about using iterators instead; it all depends what the `[]` operator actually does. You'll have to look at the rapidjson documentation to see if it performs bounds-checking, etc.

dhritzkiv replied:
I tried implementing it as an iterator, but I couldn't figure out how to do it properly with rapidjson arrays: objects and arrays had different methods for iterating, and returned different values. So I gave up. 🤷‍♂️

dhritzkiv commented on code/glTF2Asset.inl:
```diff
@@ -187,47 +194,75 @@ inline void LazyDict<T>::DetachFromDocument()
-    if (!mDict) {
-        throw DeadlyImportError("GLTF: Missing section \"" + std::string(mDictId) + "\"");
-    }
+    return Create(id);
```
Should this Create a new Resource? or should it throw? (as `::Get(unsigned int i)` will throw if the access is out-of-bounds). I think the latter

johnmaf replied:
Rather than throw, you could also return an empty `Ref` (`return Ref<T>();`), which might also allow you to get rid of the `Has` member function, since you could do something like:
```c++
if (Ref<T> ref = thing.Get(id)) {
    doSomethingWith(ref);
}
```
Does that make sense?

dhritzkiv replied:
Totally makes sense. Was trying to do exactly that for too long on Sunday, but couldn't get the syntax right.

johnmaf replied:
Gotcha. In case you're curious, the reason we can treat a `Ref` as a `bool`, as in the above `if` statement, is because there's an `operator bool()` member function defined on the class: https://github.com/Matter-and-Form/assimp/blob/26c3bbfc0ffcae368c5eb9229ee67c35875a6065/code/glTF2Asset.h#L360. It also overloads the `->` operator, so you can treat it like a pointer.

dhritzkiv commented on code/glTF2Exporter.cpp:
```diff
@@ -234,63 +223,86 @@ inline Ref<Accessor> ExportData(Asset& a, std::string& meshName, Ref<Buffer>& bu
-    // Hard coded Texture filtering options because I do not know where to find them in the aiMaterial.
-    texture->sampler->magFilter = SamplerMagFilter_Linear;
-    texture->sampler->minFilter = SamplerMinFilter_Linear_Mipmap_Linear;
+    if (mAsset->samplers.Has(id.c_str())) {
```
I had to create a new method (`Has`) to check if an object of an id already existed. Is this valid? Or could this have been accomplished using an existing method?

dhritzkiv committed:
Revert "Re-include gltf (1) importer"

This reverts commit 26c3bbfc0ffcae368c5eb9229ee67c35875a6065.

Causes an exception on export

dhritzkiv committed:
Use different form of index accessor

dhritzkiv committed:
Remove need for Has by returning an empty Ref in Get

dhritzkiv committed:
Define default material values as static constants

for reuse

johnmaf commented on code/glTF2Importer.cpp:
```diff
@@ -0,0 +1,651 @@
+                    break;
+            }
+
+            Mesh::Primitive::Attributes& attr = prim.attributes;
```
There seem to be some formatting issues below (probably tabs vs. spaces)

johnmaf commented:
Have you thought about trying to improve code re-use between the two glTF versions? It seems to me that a lot of the glTFAsset infrastructure (`Ref`s and `Asset`s and `LazyDictBase`s and whatnot) could be shared, as could the exporters and importers, potentially through sub-classing (?).

dhritzkiv committed:
Implement pbrSpecularGlossiness property as Nullable

johnmaf commented on code/glTF2Asset.h:
```diff
@@ -670,7 +669,12 @@ namespace glTF2
+	static const vec4 defaultBaseColor = {1, 1, 1, 1};
+	static const vec3 defaultEmissiveFactor = {0, 0, 0};
+	static const vec4 defaultDiffuseFactor = {1, 1, 1, 1};
+	static const vec3 defaultSpecularFactor = {1, 1, 1};
```
Whoops, mislead you about the `static` thing. `static` has 2 separate meanings in C++: if you declare a member of a `class` as `static`, it'll be allocated once during the lifetime of the program. However, if you declare a member of a `namespace` as static, it tells the compiler that that variable should use "internal linkage", meaning it's only available inside the current compilation unit (usually a .cpp file). Since these defaults are at the namespace level, it's sufficient to declare them as `const vec3`, etc. (i.e. without the `static` qualifier).

dhritzkiv replied:
Ok!

dhritzkiv committed:
Fix indentation

dhritzkiv committed:
Remove unecessary `static`

dhritzkiv commented:
> Have you thought about trying to improve code re-use between the two glTF versions? It seems to me that a lot of the glTFAsset infrastructure (Refs and Assets and LazyDictBases and whatnot) could be shared, as could the exporters and importers, potentially through sub-classing (?).

I have thought about it, but it wasn't a priority, as my focus was simply getting glTF2 working. Maybe it's something we can do after we get this merged into assimp proper.

johnmaf commented on code/glTF2AssetWriter.inl:
```diff
@@ -397,17 +454,24 @@ namespace glTF2 {
 
     inline void Write(Value& obj, Sampler& b, AssetWriter& w)
```
Because all of the properties of a `Sampler` are optional, it's possible that a totally empty `Sampler` may be written to the file, which is valid glTF, but a better solution might be to omit the sample altogether.

This is a very common case: exporting an OBJ with textures as glTF2 will result in empty `Sampler`s.

dhritzkiv replied:
Yep. I wanted to do something like that, but didn't spend too much time on it this weekend. I couldn't figure out an easy way to check if the Sampler resource was empty, and prevent if from being written: There's a general AssetWriter::WriteObjects method which delegates to each Type's Write method. But then I'm not sure how to cleanly check If the object is empty and stop it from being written. Further, I'm not sure how to prevent texture from referencing that now-deleted Sampler.

Or is there a simpler solution?

johnmaf replied:
The empty samplers are created in https://github.com/Matter-and-Form/assimp/blob/feature/gltf2-importer/code/glTF2Exporter.cpp#L243. So, you should be able to check if at least one of the sampler parameters is non-default, and only add it to the texture in that case, right?

dhritzkiv replied:
True. At that point (in the Exporter), it's not actually empty, but has default values set. It's only in AssetWriter that those default values get pruned.

johnmaf commented:
Transparencies are ignored when exporting glTF2 models with textures. In particular, exporting the [Earth model](https://github.com/Matter-and-Form/assimp/files/1278989/earth.zip) as glTF2 causes the dissolve of .8 on the atmosphere material to be ignored, meaning the earth material isn't rendered.

This could be implemented by combining the baseColorFactor and baseColorTexture properties of the pbrMetallicRoughness object (the Factor is to be multiplied by the Texture values by the renderer).

johnmaf commented on code/glTF2Importer.cpp:
```diff
@@ -0,0 +1,651 @@
+    return &desc;
+}
+
+bool glTF2Importer::CanRead(const std::string& pFile, IOSystem* pIOHandler, bool checkSig) const
```
This should check for an `"asset": { "version": "2.*" }` before returning `true`. Not sure what the best way to do that would be; any thoughts?

dhritzkiv replied:
None at the moment. I would imagine the entire JSON doc would need to be loaded into memory and parsed to determine the version. Or maybe a regex on the input stream:

```js
//key of "version"
//possible whitespace before and after the colon character
//value of 2 (with possible minor version) as a string
/"version"(\s)*:(\s)*"2(\.[0-9]+)?"/
```

But I haven't worked with Assimp's IO yet, so I'm not sure.

johnmaf replied:
The `BaseImporter` class from which all the importers derive has a `SearchFileHeaderForToken` function that may be useful, if only as a template to implement something smarter. Really, though, I don't think it would be the worst thing in the world to just parse the JSON and check the `"asset"` that way, and it'd be much less error-prone.

dhritzkiv replied:
How would you feel about taking a stab at that?

johnmaf replied:
Sure! Do you forsee any issues with parsing the JSON and checking for the asset version? That'd be my preferred method.

dhritzkiv replied:
Nope. Only that it might be slower than checking the file's string against a regex quickly. But it'd probably be negligibly slower, and the regex isn't all that accurate.

dhritzkiv committed:
Define gltf material property names as constants

dhritzkiv committed:
Fix METALLIC_FACTOR typo

dhritzkiv committed:
Set the metallicFactor to 0 if source file doesn’t have metallicFactor

Otherwise, the models have full metallic materials and appear very dim

dhritzkiv committed:
Set alphaMode, baseColorFactor opacity when model’s opacity isn’t 1

dhritzkiv committed:
Remove commented out code

johnmaf committed:
Restrict search for OFF header to first 3 bytes

johnmaf committed:
Add support for importing both glTF and glTF2 files

johnmaf committed:
Fix parsing of glTF version

johnmaf committed:
Merge branch 'feature/gltf2-importer' of github.com:Matter-and-Form/assimp into feature/gltf2-importer

dhritzkiv commented on code/glTF2Importer.cpp:
```diff
@@ -100,24 +102,19 @@ const aiImporterDesc* glTF2Importer::GetInfo() const
+    if (checkSig && pIOHandler) {
+        glTF2::Asset asset(pIOHandler);
+        try {
+            asset.Load(pFile, extension == "glb");
```
Should this line be in here?

dhritzkiv commented on code/glTFAssetWriter.inl:
```diff
@@ -606,13 +606,8 @@ namespace glTF {
         Value asset;
         asset.SetObject();
-        {
-            char versionChar[10];
```
Out of curiosity, what was happening in this block?

dhritzkiv committed:
Handle version as int in gltf

johnmaf committed:
Fix format specifiers in glTF version parser

dhritzkiv committed:
Remove premultipliedAlpha from gltf2

dhritzkiv committed:
Load gltf .bin files from correct directory

dhritzkiv committed:
more specific token search for Collada Loader

collada can appear in many files, such as glTFs via the “generator” field (in the form of collada2gltf)

dhritzkiv committed:
Add gltf2 basic unit test

dhritzkiv committed:
Add glTF2Export test
